src/signature.rs

```diff
@@ -164,136 +171,295 @@ impl<const N: usize> Signature<N> {
+                            {
+                                cursor +=
+                                    skip_offsets[*scan.add(N.saturating_sub(1)) as usize] as usize;
+                                continue 'outer;
+                            } else {
+                                mask = mask.add(8);
+                                needle = needle.add(8);
+                                i += 8;
+                            }
+                        }
+
+                        while i + 4 <= N {
+                            if scan.add(i).cast::<u32>().read_unaligned()
+                                & mask.cast::<u32>().read_unaligned()
+                                != needle.cast::<u32>().read_unaligned()
+                            {
+                                cursor +=
+                                    skip_offsets[*scan.add(N.saturating_sub(1)) as usize] as usize;
+                                continue 'outer;
+                            } else {
+                                mask = mask.add(4);
+                                needle = needle.add(4);
+                                i += 4;
+                            }
+                        }
+
+                        while i + 2 <= N {
+                            if scan.add(i).cast::<u16>().read_unaligned()
+                                & mask.cast::<u16>().read_unaligned()
+                                != needle.cast::<u16>().read_unaligned()
+                            {
+                                cursor +=
+                                    skip_offsets[*scan.add(N.saturating_sub(1)) as usize] as usize;
+                                continue 'outer;
+                            } else {
+                                mask = mask.add(2);
+                                needle = needle.add(2);
+                                i += 2;
+                            }
+                        }
+
+                        while i < N {
+                            if *scan.add(i) & *mask != *needle {
+                                cursor +=
+                                    skip_offsets[*scan.add(N.saturating_sub(1)) as usize] as usize;
+                                continue 'outer;
+                            } else {
+                                mask = mask.add(1);
+                                needle = needle.add(1);
+                                i += 1;
+                            }
+                        }
+
+                        let current_cursor = cursor;
+                        cursor = cursor + 1;
+                        return Some(current_cursor);
                     }
-                    let offset = skip_offsets[scan[end] as usize];
-                    current += offset as usize;
                 }
-                None
             }
-        }
+        })
+        .fuse()
     }
+}
+
+/// Trait that provides scanning methods for the `Signature` type.
+pub trait SignatureScanner {
```

Comment:
Why is this a trait?

Reply:
I thought having a trait could allow for implementation of custom signature scanning functions. But if you find this unnecessary, this can be removed.

Reply:
I think we only need the trait whenever someone actually abstracts over it (if we have multiple in asr for example)